File: skel/examples/example_water_spsp.c
#include <stdio.h>

#include "simint/simint.h"

/* Quick and dirty example of using Simint
 *
 * We use hard-coded water/STO-3G as a simple example
 */
int main(int argc, char ** argv)
{
    ////////////////////////////////
    // General workflow as follows
    ////////////////////////////////


    ////////////////////////////////////////
    // 1. Initialize the library
    simint_init();


    ////////////////////////////////////////
    // 2. Create your shells
    struct simint_shell s_shells[4];
    struct simint_shell p_shells[1];

    // initialize the shells
    simint_initialize_shell(&s_shells[0]);
    simint_initialize_shell(&s_shells[1]);
    simint_initialize_shell(&s_shells[2]);
    simint_initialize_shell(&s_shells[3]);
    simint_initialize_shell(&p_shells[0]);


    // allocate the memory
    // arguments to simint_allocate_shell
    //    Number of primitives
    //    Pointer to the shell
    // Note that this doesn't actually set the
    // nprim member of the structure
    simint_allocate_shell(3, &s_shells[0]);
    simint_allocate_shell(3, &s_shells[1]);
    simint_allocate_shell(3, &s_shells[2]);
    simint_allocate_shell(3, &s_shells[3]);
    simint_allocate_shell(3, &p_shells[0]);

    //
    // Coordinates are in atomic units!
    //
    // oxygen
    s_shells[0].am = 0;  s_shells[0].nprim = 1;
    s_shells[0].x = 0.00000; s_shells[0].y = 0.0; s_shells[0].z = -0.02;
    s_shells[0].alpha[0] = 1.;  s_shells[0].coef[0] =  1.;

    s_shells[1].am = 0;  s_shells[1].nprim = 1;
    s_shells[1].x = 0.00000; s_shells[1].y = 0.0; s_shells[1].z = -0.02;
    s_shells[1].alpha[0] =   0.4;  s_shells[1].coef[0] = 1.;

    p_shells[0].am = 1;  p_shells[0].nprim = 1;
    p_shells[0].x = 0.00000; p_shells[0].y = 0.0; p_shells[0].z = -0.02;
    p_shells[0].alpha[0] =   0.4;  p_shells[0].coef[0] =  1.;

    // hydrogen 1
    s_shells[2].am = 0;  s_shells[2].nprim = 1;
    s_shells[2].x = -0.74; s_shells[2].y = 0.0; s_shells[2].z = -0.76;
    s_shells[2].alpha[0] =  0.5;  s_shells[2].coef[0] =  1.;

    // hydrogen 2
    s_shells[3].am = 0;  s_shells[3].nprim = 1;
    s_shells[3].x =  0.74; s_shells[3].y = 0.0; s_shells[3].z = -0.76;
    s_shells[3].alpha[0] = 0.5;  s_shells[3].coef[0] =  1.;


    ////////////////////////////////////////
    // 3. Normalize
    // Arguments to simint_normalize_shells:
    //   number of shells
    //   pointer to array of shells
    simint_normalize_shells(4, s_shells);
    simint_normalize_shells(1, p_shells);


    ////////////////////////////////////////
    // 4. Create your multishell pairs
    // Could obviously be done on the fly if needed
    struct simint_multi_shellpair s1p_pair;
    struct simint_multi_shellpair s2p_pair;
    simint_initialize_multi_shellpair(&s1p_pair);
    simint_initialize_multi_shellpair(&s2p_pair);

    // Last argument - set to one of the following to enable primitive screening
    //       1 - Schwarz screening
    //       2 - Fast Schwarz screening
    simint_create_multi_shellpair(1, &s_shells[3], 1, p_shells, &s2p_pair, 0);
    simint_create_multi_shellpair(1, &s_shells[2], 1, p_shells, &s1p_pair, 0);


    ////////////////////////////////////////
    // 5. Compute your integrals
    // For this example, I'll just allocate
    // space for all (although this overestimates
    // since we are exploiting some symmetry)

    // Allocate some required workspace
    // Arguments to simint_ostei_workmem: Derivative order, max am
    double * work = SIMINT_ALLOC(simint_ostei_workmem(0, 1));

    // Argument is the number of doubles
    double * targets = SIMINT_ALLOC(7*7*7*7 * sizeof(double));
    int ncomputed = 0;
    int ntotal = 0;

    // 3rd argument is the screening tolerance. Set to 0.0 to disable
    ncomputed = simint_compute_eri(&s2p_pair, &s1p_pair, 0, work, targets+ntotal);
    printf("Computed %d contracted s2ps1p integrals\n", ncomputed);
    ntotal += ncomputed;
    ntotal = 9;

    printf("++Computed %d contracted integrals\n", ntotal);

    for(int i = 0; i < ntotal; i++)
        printf(" %d  %20.6f\n", i+1, targets[i]);

    ////////////////////////////////////////
    // 6. Clean up and finalize the library
    //
    simint_free_shell(&s_shells[0]);
    simint_free_shell(&s_shells[1]);
    simint_free_shell(&s_shells[2]);
    simint_free_shell(&s_shells[3]);
    simint_free_shell(&p_shells[0]);
    simint_free_multi_shellpair(&s1p_pair);
    simint_free_multi_shellpair(&s2p_pair);
    SIMINT_FREE(targets);
    SIMINT_FREE(work);
    simint_finalize();

    return 0;
}

File: skel/examples/example_water_spsp.f90
program SimintTest
use SimintFortran
use iso_c_binding
implicit none
type(c_simint_shell), target :: s_shell(4),p_shell(1)
type(c_simint_multi_shellpair), target :: s2p_msh, s1p_msh

double precision :: alpha(1), coef(1)
double precision :: integrals(1000)
double precision, allocatable :: work(:)

double precision, pointer :: p1(:), p2(:)
integer :: i, ncomputed
integer :: worksize

call simint_init()


worksize = simint_eri_worksize(0, 3)
allocate(work(worksize))



call simint_initialize_shell(s_shell(1))
call simint_initialize_shell(s_shell(2))
call simint_initialize_shell(s_shell(3))
call simint_initialize_shell(s_shell(4))
call simint_initialize_shell(p_shell(1))
call simint_initialize_multi_shellpair(s2p_msh)
call simint_initialize_multi_shellpair(s1p_msh)
! oxygen s
alpha(1)=1d0
coef(1) =1d0
call simint_create_shell(1, 0, 0.0d0, 0.0d0, -0.02d0, &
                          alpha(1), coef(1), s_shell(1)) 
! oxygen sp
alpha(1)=0.4d0
coef(1) =1d0
call simint_create_shell(1, 0, 0.0d0, 0.0d0, -0.02d0, &
                         alpha(1), coef(1), s_shell(2)) 
call simint_create_shell(1, 1, 0.0d0, 0.0d0, -0.02d0, &
                        alpha(1), coef(1), p_shell(1)) 
!hydrogen s
alpha(1)=0.5d0
coef(1) =1d0
call simint_create_shell(1, 0, -0.74d0, 0.0d0, -0.76d0, &
                         alpha(1), coef(1), s_shell(3)) 
call simint_create_shell(1, 0,  0.74d0, 0.0d0, -0.76d0, &
                         alpha(1), coef(1), s_shell(4)) 

call C_F_POINTER(s_shell(1)%alpha, p1, shape=[s_shell(1)%nprim])
call C_F_POINTER(s_shell(1)%coef, p2, shape=[s_shell(1)%nprim])
write(*,*) "s_shell1  info"
do i = 1, s_shell(1)%nprim
  write(*,*) p1(i), p2(i)
end do
call C_F_POINTER(s_shell(2)%alpha, p1, shape=[s_shell(2)%nprim])
call C_F_POINTER(s_shell(2)%coef, p2, shape=[s_shell(2)%nprim])
write(*,*) "s_shell1  info"
do i = 1, s_shell(2)%nprim
  write(*,*) p1(i), p2(i)
end do
call C_F_POINTER(p_shell(1)%alpha, p1, shape=[p_shell(1)%nprim])
call C_F_POINTER(p_shell(1)%coef, p2, shape=[p_shell(1)%nprim])
write(*,*) "p_shell  info"
do i = 1, s_shell(1)%nprim
  write(*,*) p1(i), p2(i)
end do

call simint_normalize_shells(4, s_shell)
call simint_normalize_shells(1, p_shell(1))

call simint_create_multi_shellpair(1, s_shell(4), 1, p_shell(1), s2p_msh, 0)
call simint_create_multi_shellpair(1, s_shell(3), 1, p_shell(1), s1p_msh, 0)

call C_F_POINTER(s1p_msh%alpha, p1, shape=[s1p_msh%nprim])
call C_F_POINTER(s1p_msh%prefac, p2, shape=[s1p_msh%nprim])
write(*,*) "s2p Shell Pair info"
do i = 1, s1p_msh%nprim
  write(*,*) p1(i), p2(i)
end do

call C_F_POINTER(s2p_msh%alpha, p1, shape=[s2p_msh%nprim])
call C_F_POINTER(s2p_msh%prefac, p2, shape=[s2p_msh%nprim])
write(*,*) "s2p Shell Pair info"
do i = 1, s2p_msh%nprim
  write(*,*) p1(i), p2(i)
end do


ncomputed = simint_compute_eri(s2p_msh, s1p_msh, 0.0d0, work, integrals)
write(6,*) ' ncomputed ',ncomputed
do i=1,9
   write(*,'(i4,f20.6)') i,integrals(i)
enddo

deallocate(work)
call simint_free_shell(s_shell(1))
call simint_free_shell(s_shell(2))
call simint_free_shell(s_shell(3))
call simint_free_shell(s_shell(4))
call simint_free_shell(p_shell(1))
call simint_free_multi_shellpair(s2p_msh)
call simint_free_multi_shellpair(s1p_msh)

call simint_finalize()

end program
